javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll

```diff
@@ -1347,6 +1347,11 @@ module DataFlow {
       succ = lvalueNode(pattern.getValuePattern())
     )
     or
+    exists(PropertyPattern pattern | 
+      pred = TPropNode(pattern) and
+      succ = lvalueNode(pattern.getValuePattern().(ObjectPattern).getRest())
+    )
```

Comment:
Let's just review the runtime behavior of rest patterns:
- `{ ...args } = x` stores a shallow copy of `x` in `args`.
- `{ p1, p2, ...args } = x` stores a shallow copy of `x` _without_ properties `p1`, `p2`, etc in `args`.

What this change is doing is essentially modelling this copying step via a pure data flow step `x -> args`.

There are two issues here:
- How do we best express that step?
- Is it safe to use a pure data flow step for this?

For the first issue, I think the step should go from the object pattern itself to its rest pattern:
```codeql
exists(ObjectPattern pattern |
  pred = lvalueNode(pattern) and
  succ = lvalueNode(pattern.getRest())
)
```

For the second issue, I can completely see how this would work great most of the time, but not always. For example, this would cause `getALocalSource` to backtrack through a copy step, making some of our store-steps spuriously propagate back to the original object.

I think the best would be to add it as an `AdditionalFlowStep`. Those steps don't affect `getALocalSource` but do affect all data-flow configurations. And since the above formulation doesn't depend on any internal predicates it should be a straightforward addition.

Reply:
Thanks. 

The only data-flow related use of rest-patterns I could fine was in `DefUse::lvalAux`, I think that threw me off. 

I've added it as a default taint-tracking step for now.